src/v/ssx/include/ssx/sharded_ptr.h

```diff
@@ -84,6 +81,40 @@ class sharded_ptr {
         return reset(std::make_shared<T>(std::forward<Args>(args)...));
     }
 
+    /// updates the managed object u with an update function
+    /// The update function takes a copy of the underlying object and creates a
+    /// new object using the update function. The update is executed inside the
+    /// guard of the update mutex to ensure that updates are executed
+    /// sequentially.
+    /// Must be called on the home shard and is safe to call concurrently.
+    /// returns an ss::broken_semaphore if stop() has been called.
+    template<typename F>
+    requires std::is_invocable_r_v<T, F, T>
+    ss::future<> update(F&& update_func) {
+        return update_shared([update_func{std::forward<F>(update_func)}](
+                               const std::shared_ptr<T>& v) {
+            return std::make_shared<T>(update_func(v ? *v : T{}));
```

Comment:
The `update_func` should not be allowed to take a non-const reference to the state.
```suggestion
            return std::make_shared<T>(
              update_func(v ? std::as_const(*v) : T{}));
```